File: findviz/static/js/upload/file.js
// file.js
import { 
  FILE_UPLOAD_FIELDS,
  FMRI_FILE_TYPES, 
  UPLOAD_ENDPOINTS, 
} from './constants.js';
import UploadErrorHandler from './error.js';

/** 
 * @typedef {import('./types').UploadResponse} UploadResponse
 * @typedef {import('./types').NiftiFiles} NiftiFiles 
 * @typedef {import('./types').GiftiFiles} GiftiFiles
 * @typedef {import('./types').TimeSeriesFile} TimeSeriesFile
 * @typedef {import('./types').TaskDesignFile} TaskDesignFile
 * @typedef {import('./types').UploadFormData} UploadFormData
 */


/**
 * Class for handling file uploads from findviz upload modal. Files 
 * and form information are passed to Flask routes for validation and
 * server-side processing.
 */
class FileUploader {
  /**
   * Construct a FileUpload class with a callback function (uploadComplete)
   * to be executed on successful upload
   * @param {{ (data: object, plotType: string)  }} onUploadComplete 
   */
  constructor(onUploadComplete) {
    // counter variable to generate unique ids for time series switches
    this.tsFileCounter = 0
    // callback function to run after upload completion
    this.onUploadComplete = onUploadComplete;
    // initialize upload error handler
    this.errorHandler = new UploadErrorHandler();
    // Set up form submission listener
    document.getElementById('upload-form').onsubmit = async (event) => {
      // prevent page reload
      event.preventDefault();
      // get active tabe in form submission
      let activeTab = document.querySelector('.nav-pills .active').getAttribute('href'); // Get the active tab
      // Check whether the input is nifti or gifti based on active tab
      let fmriFileType
      if (activeTab == '#nifti') {
        fmriFileType = FMRI_FILE_TYPES.NIFTI;
      }
      else if (activeTab == '#gifti') {
        fmriFileType = FMRI_FILE_TYPES.GIFTI;;
      }
      this.uploadFiles(event, fmriFileType);
    };

    // get error message placeholder
    this.errorMessageDiv = document.getElementById('error-message-upload');

    // Initialize modal listeners
    this.initializeModalListeners();

    // Initialize TR input form listeners
    this.initializeTRFormListeners();
  }

  /**
   * Upload files to server for validation and processing
   * @param {Event} event - Form submission event
   * @param {string} fmriFileType - Type of fMRI files ('nifti' or 'gifti')
   * @returns {Promise<UploadResponse>}
   */

  async uploadFiles(event, fmriFileType) {
    this.errorHandler.showSpinner();

    try {
        const uploadData = this.getFiles();
        uploadData.append('fmri_file_type', fmriFileType);

        const response = await fetch(UPLOAD_ENDPOINTS.FILES, {
            method: 'POST',
            body: uploadData
        });

        if (response.ok) {
            const data = await response.json();
            document.getElementById('fmri-visualization-container').style.display = 'block';
            this.onUploadComplete(data, fmriFileType);
            $('#upload-modal').modal('hide');
        } else {
            await this.errorHandler.handleServerError(response, fmriFileType);
        }
    } catch (error) {
        console.error('Error during file upload:', error);
        this.errorHandler.showServerErrorModal();
    } finally {
        this.errorHandler.hideSpinner();
    }
  }

  /**
   * Upload scene file and process cached data
   * @param {object} data - Scene file data
   * @returns {Promise<void>}
   */
  async uploadSceneFile(event, data) {
    this.errorHandler.showSpinner();

    try {
        const sceneFile = event.target.files[0];
        const formData = new FormData();
        formData.append('scene_file', sceneFile);

        const response = await fetch(UPLOAD_ENDPOINTS.SCENE, {
            method: 'POST',
            body: formData
        });

        if (!response.ok) {
            this.errorHandler.showSceneErrorModal();
            throw new Error('Failed server processing of uploaded cache');
        }

        const data = await response.json();
        // get data and call callback function
        // assign time series output to data
        if (data.ts_enabled) {
          Object.assign(data, {timeCourses: data.timeseries})
        } else {
          // If no time courses append empty lists
          Object.assign(data, {timeCourses: {ts: [], tsLabels: []}})
        }

        if (data.task_enabled) {
          Object.assign(data, {taskConditions: data.task})
        } else {
          Object.assign(data, {taskConditions: null})
        }
        this.onUploadComplete(data, data.file_type);
        $('#upload-modal').modal('hide');
    } catch (error) {
        console.error('Error during scene file upload:', error);
        this.errorHandler.showServerErrorModal();
    } finally {
        this.errorHandler.hideSpinner();
    }
  }

  /**
   * Add new time series file input fields to the form
   * Creates a new row with file input, header switch, label textarea, and remove button
   */
  addTimeSeriesFile() {
    const container = document.getElementById('time-series-container');
    const filePair = document.createElement('div');

    // increase counter
    this.tsFileCounter += 1
    // create unique id for switch input
    const uniqueID = `${FILE_UPLOAD_FIELDS.TS.HEADER}-${this.tsFileCounter}`;

    filePair.className = 'times-series-file-pair row mb-2';
    filePair.innerHTML = `
      <div class="col-6">
        <span class="d-inline-block text-secondary">Time Series File (.txt, .csv, optional)</span>
        <i class="fa-solid fa-triangle-exclamation ${FILE_UPLOAD_FIELDS.TS.FILE}-error" style="color: #e93407; display: none;"></i>
        <input type="file" class="form-control-file ${FILE_UPLOAD_FIELDS.TS.FILE} pt-2" data-index="${this.tsFileCounter - 1}">
      </div>
      <div class="col-4">
        <div class="custom-control custom-switch">
          <input type="checkbox" class="custom-control-input ${FILE_UPLOAD_FIELDS.TS.HEADER}" id="${uniqueID}">
          <label class="custom-control-label" for="${uniqueID}">Header</label>
          <span class="fa fa-info-circle ml-1 toggle-immediate" data-toggle="tooltip" data-placement="top" title="Does the file have a header (i.e. name) in the first row?" aria-hidden="true"></span>
        </div>
        <textarea class="form-control ${FILE_UPLOAD_FIELDS.TS.LABEL}" placeholder="Label" rows="1"></textarea>
      </div>
      <div class="col-2 mt-4">
        <button type="button" class="remove-time-series btn btn-danger btn-sm">x</button>
      </div>
    `;
    // Append the new file input to the container
    container.appendChild(filePair);

    // Enable the tooltip
    const tooltip = filePair.querySelector('.toggle-immediate');
    $(tooltip).tooltip()
    // Select the file input we just added
    const newFileInput = filePair.querySelector(`.${FILE_UPLOAD_FIELDS.TS.FILE}`);

     // Select the header switch we just added and add the event listener
    const headerSwitch = filePair.querySelector('.custom-control-input');
    const labelTextarea = filePair.querySelector(`.${FILE_UPLOAD_FIELDS.TS.LABEL}`);

    headerSwitch.addEventListener('change', () => {
      if (headerSwitch.checked) {
        const newFileIndex = newFileInput.dataset.index;
        this.handleHeaderSwitch(newFileInput, labelTextarea, newFileIndex);
      } else {
        // Clear label if switch is toggled off
        labelTextarea.value = '';
      }
    });

    // Select the remove button we just added and add the event listener
    const removeButton = filePair.querySelector('button');
    removeButton.addEventListener('click', () => {
      this.removeTimeSeriesFile(removeButton);
    });
  }

  /**
   * Clear all fMRI file inputs for a specific type
   * @param {string} fmriType - Type of fMRI files to clear ('nifti' or 'gifti')
   */
  clearfMRIFiles(fmriType) {
    // Get nifti files
    if (fmriType == FMRI_FILE_TYPES.NIFTI) {
      document.getElementById(FILE_UPLOAD_FIELDS.FMRI.NIFTI.FUNC).value = '';
      document.getElementById(FILE_UPLOAD_FIELDS.FMRI.NIFTI.ANAT).value = '';
      document.getElementById(FILE_UPLOAD_FIELDS.FMRI.NIFTI.MASK).value = '';
    // remove gifti files
    } else if (fmriType == FMRI_FILE_TYPES.GIFTI) {
      document.getElementById(FILE_UPLOAD_FIELDS.FMRI.GIFTI.LEFT_FUNC).value = '';
      document.getElementById(FILE_UPLOAD_FIELDS.FMRI.GIFTI.RIGHT_FUNC).value = '';
      document.getElementById(FILE_UPLOAD_FIELDS.FMRI.GIFTI.LEFT_MESH).value = '';
      document.getElementById(FILE_UPLOAD_FIELDS.FMRI.GIFTI.RIGHT_MESH).value = '';
    }
  }

  /**
   * Get NIFTI file inputs and add to form data
   * @returns {FormData & NiftiFiles}
   * @private
  */
    _getNiftiFiles() {
      const formData = new FormData();
      formData.append(
          'nii_func', 
          document.getElementById(FILE_UPLOAD_FIELDS.FMRI.NIFTI.FUNC).files[0]
      );
      formData.append(
          'nii_anat', 
          document.getElementById(FILE_UPLOAD_FIELDS.FMRI.NIFTI.ANAT).files[0]
      );
      formData.append(
          'nii_mask', 
          document.getElementById(FILE_UPLOAD_FIELDS.FMRI.NIFTI.MASK).files[0]
      );
      return formData;
  }

  /**
   * Get GIFTI file inputs and add to form data
   * @returns {FormData & GiftiFiles}
   * @private
   */
  _getGiftiFiles() {
      const formData = new FormData();
      formData.append(
          'left_gii_func', 
          document.getElementById(FILE_UPLOAD_FIELDS.FMRI.GIFTI.LEFT_FUNC).files[0]
      );
      formData.append(
          'right_gii_func',
          document.getElementById(FILE_UPLOAD_FIELDS.FMRI.GIFTI.RIGHT_FUNC).files[0]
      );
      formData.append(
          'left_gii_mesh',
          document.getElementById(FILE_UPLOAD_FIELDS.FMRI.GIFTI.LEFT_MESH).files[0]
      );
      formData.append(
          'right_gii_mesh',
          document.getElementById(FILE_UPLOAD_FIELDS.FMRI.GIFTI.RIGHT_MESH).files[0]
      );
      return formData;
  }

  /**
   * Get time series file inputs and add to form data
   * @returns {{formData: FormData, hasTimeSeries: boolean, files: TimeSeriesFile[]}}
   * @private
   */
  _getTimeSeriesFiles() {
      const formData = new FormData();
      const timeSeriesArray = [];
      const timeSeriesLabelsArray = [];
      const timeSeriesHeaderArray = [];

      const timeSeriesInput = document.querySelectorAll(`.${FILE_UPLOAD_FIELDS.TS.FILE}`);
      const timeSeriesInputLabel = document.querySelectorAll(`.${FILE_UPLOAD_FIELDS.TS.LABEL}`);
      const timeSeriesHeader = document.querySelectorAll(`.${FILE_UPLOAD_FIELDS.TS.HEADER}`);

      for (const [index, ts] of timeSeriesInput.entries()) {
          if (ts.files.length > 0) {
              timeSeriesArray.push(ts.files[0]);
              let tsLabel = timeSeriesInputLabel[index].value || ts.files[0].name;
              timeSeriesLabelsArray.push(tsLabel);
              timeSeriesHeaderArray.push(timeSeriesHeader[index].checked);
          }
      }

      formData.append('ts_files', timeSeriesArray);
      formData.append('ts_labels', timeSeriesLabelsArray);
      formData.append('ts_headers', timeSeriesHeaderArray);

      return {
          formData,
          hasTimeSeries: timeSeriesArray.length > 0
      };
  }

  /**
   * Get task design file inputs and add to form data
   * @returns {{formData: FormData, hasTaskDesign: boolean, taskDesign: TaskDesignFile|null}}
   * @private
   */
  _getTaskDesignFiles() {
      const formData = new FormData();
      const taskInput = document.getElementById(FILE_UPLOAD_FIELDS.TASK.FILE).files[0];

      formData.append('task_file', taskInput);
      formData.append(
          'tr', 
          document.getElementById(FILE_UPLOAD_FIELDS.TASK.TR).value
      );
      formData.append(
          'slicetime_ref',
          document.getElementById(FILE_UPLOAD_FIELDS.TASK.SLICETIME).value
      );

      return {
          formData,
          hasTaskDesign: taskInput !== undefined
      };
  }

  /**
   * Get all files and form data from the upload modal
   * @returns {FormData & UploadFormData}
   */
  getFiles() {
    const masterFormData = new FormData();
    
    // Get active tab to determine fMRI file type
    const activeTab = document.querySelector('.nav-pills .active').getAttribute('href');
    const fmriFileType = activeTab === '#nifti' ? FMRI_FILE_TYPES.NIFTI : FMRI_FILE_TYPES.GIFTI;
    
    // Get all file data
    const fmriData = fmriFileType === FMRI_FILE_TYPES.NIFTI 
        ? this._getNiftiFiles() 
        : this._getGiftiFiles();
    const { formData: tsData, hasTimeSeries, files: tsFiles } = this._getTimeSeriesFiles();
    const { formData: taskData, hasTaskDesign, taskDesign } = this._getTaskDesignFiles();

    // Combine all FormData objects
    [fmriData, tsData, taskData].forEach(formData => {
        for (const [key, value] of formData.entries()) {
            masterFormData.append(key, value);
        }
    });

    // Add flags
    masterFormData.append('fmri_file_type', fmriFileType);
    masterFormData.append('ts_input', hasTimeSeries);
    masterFormData.append('task_input', hasTaskDesign);

    return masterFormData;
  }

  /**
   * Handle header switch toggle for time series files
   * Fetches and displays header from file when switch is toggled on
   * @param {HTMLInputElement} fileInput - File input element
   * @param {HTMLTextAreaElement} labelTextarea - Label textarea element
   * @param {number} fileIndex - index of file input
   * @returns {Promise<void>}
   */
  async handleHeaderSwitch(fileInput, labelTextarea, fileIndex) {
    if (!fileInput.files.length) {
        this.errorHandler.displayError('Please select a file before enabling header');
        return;
    }

    const formData = new FormData();
    formData.append('ts_file', fileInput.files[0]);
    formData.append('file_index', fileIndex);

    try {
        const response = await fetch(
            UPLOAD_ENDPOINTS.HEADER, 
            { method: 'POST', body: formData }
        );

        if (response.ok) {
            const data = await response.json();
            labelTextarea.value = data.header;
        } else {
            await this.errorHandler.handleServerError(response, 'timecourse');
        }
    } catch (error) {
      console.error('Error reading time course file to get header:', error);
      this.errorHandler.showServerErrorModal();
    }
  }

  /**
   * Initialize modal event listeners
   * Sets up listeners for tab switching, scene file upload, and time series file management
   */
  initializeModalListeners() {
    // create reference to self
    const self = this;
    // Clear inputs and error messages on tab switch
    document.querySelectorAll('.nav-pills .nav-link').forEach(tab => {
      tab.addEventListener('click', function () {
        document.getElementById('error-message-upload').style.display = 'none'; // Clear error message on tab switch
        // Get the currently active tab
        let activeTab = document.querySelector('.nav-pills .active').getAttribute('href');
        let fileType
        if (activeTab == '#nifti') {
          fileType = 'nifti';
        }
        else if (activeTab == '#gifti') {
          fileType = 'gifti';
        }
        // Reset only the FMRI inputs (Nifti or Gifti based on active tab)
        self.clearfMRIFiles(fileType);
      });
    });

    // Open file dialog when user clicks upload scene
    const uploadSceneButton = $('#uploadScene');
    const sceneFileDiv = $('#fileScene');
    uploadSceneButton.on('click', () => {
      // open file dialog
      sceneFileDiv.click();
    });

    // on scene file upload pass to flask route
    sceneFileDiv.on('change', (event) => {
      this.uploadSceneFile();
    });

    // Add event listener for bootstrap model close to clear error message
    $('#uploadModal').on('hidden.bs.modal', function (e) {
        document.getElementById('error-message-upload').style.display = 'none';
    });

    // Event listener for add physio file
    let addTSDiv = document.getElementById('add-time-series')
    addTSDiv.addEventListener('click', this.addTimeSeriesFile.bind(this))

    // On document ready add one time series file
    $(document).ready(function() {
        self.addTimeSeriesFile()
    });
  }

  /**
   * Initialize TR form input synchronization
   * Sets up listeners to keep TR values synchronized across different inputs
   */
  initializeTRFormListeners() {
    // Get references to the input fields for TRs
    // TR input field in modal for Task Design File
    const TRInput1 = document.getElementById(FILE_UPLOAD_FIELDS.TASK.TR);
    // TR input field for fmri preprocessing
    const TRInput2 = document.getElementById('filter-tr');
    // TR input field for time course preprocessing
    const TRInput3 = document.getElementById('ts-filter-tr');

    // Function to update all other TR fields
    function synchronizeInput(sourceInput, targetInputs) {
        sourceInput.addEventListener('input', function() {
            targetInputs.forEach(input => {
                input.value = sourceInput.value;
            });
        });
    }
    // Synchronize all TR input fields fields
    synchronizeInput(TRInput1, [TRInput2, TRInput3]);
    synchronizeInput(TRInput2, [TRInput1, TRInput3]);
    synchronizeInput(TRInput3, [TRInput1, TRInput2]);
  }

  /**
   * Remove time series file input row
   * @param {HTMLButtonElement} button - Remove button that was clicked
   */
  // physio file remove
  removeTimeSeriesFile(button) {
    const filePair = button.closest('.times-series-file-pair');
    filePair.remove();
    // decrease file counter
    this.tsFileCounter -= 1
  }

}


export default FileUploader;

File: findviz/routes/file.py
"""
Routes for file upload and validation
"""
from findviz.logger_config import setup_logger
from flask import Blueprint, request, make_response, jsonify
from findviz.routes.shared import data_manager
from findviz.routes.utils import convert_value, Routes
from findviz.viz import exception
from findviz.viz.io.cache import Cache
from findviz.viz.io.nifti import NiftiFiles
from findviz.viz.io.gifti import GiftiFiles
from findviz.viz.io.timecourse import get_ts_header
from findviz.viz.io.upload import FileUpload

# Get logger instance for this module
logger = setup_logger(__name__)

# create blueprint
file_bp = Blueprint('file', __name__)


@file_bp.route(Routes.CHECK_CACHE.value, methods=['GET'])
def check_cache():
    """Check if cache exists and return cached data if it does"""
    logger.info("Checking cache status")
    cache = Cache()
    has_cache = cache.exists()
    logger.info(f"Cache check: exists={has_cache}, path={cache.get_cache_path()}")
    if has_cache:
        try:
            cached_data = cache.load()
            # Determine plot type from cached data
            logger.info("Cache found and loaded successfully")
            return jsonify({
                'has_cache': True,
                'cache_data': cached_data,
                'plot_type': cached_data['file_type']
            })
        except Exception as e:
            logger.error("Error loading cached data: %s", str(e), exc_info=True)
            return jsonify({
                'has_cache': False,
                'error': str(e)
            })
    
    return jsonify({'has_cache': False})


@file_bp.route(Routes.CLEAR_CACHE.value, methods=['POST'])
def clear_cache():
    """Clear the cache"""
    logger.info("Clearing cache")
    try:
        cache = Cache()
        cache.clear()
        return jsonify({'success': True})
    except Exception as e:
        logger.error("Error clearing cache: %s", str(e), exc_info=True)
        return jsonify({'success': False, 'error': str(e)})

# Get header of time course file
@file_bp.route(Routes.GET_HEADER.value, methods=['POST'])
def get_header():
    # get time course file and its index
    ts_file = request.files.get('ts_file')
    ts_index = convert_value(request.form.get('file_index'))
    try:
        header = get_ts_header(ts_file, ts_index)
        logger.info("Successfully extracted header from file: %s", ts_file.filename)
    except (
        exception.FileInputError,
        exception.FileUploadError
    ) as e:
        logger.error("Error reading time series file header: %s", 
                        str(e), exc_info=True)
        return make_response({
            "error": e.message,
            "file_type": e.file_type,
            "fields": e.field,
            "index": e.index
        }, 400)
    except Exception as e:
        logger.critical("Unexpected error reading time series file header: %s", 
                        str(e), exc_info=True)
        return make_response({
            "error": "An unexpected error occurred while reading the time series file header",
            "file_type": None,
            "fields": None,
            "index": None
        }, 500)

    return make_response({
            "header": header
        }, 201)

# Upload file input
@file_bp.route(Routes.UPLOAD_FILES.value, methods=['POST'])
def upload():
    logger.info("Starting file upload process")
    # get form parameters
    fmri_file_type = request.form.get('fmri_file_type')
    timecourse_input = convert_value(request.form.get('ts_input'))
    task_input = convert_value(request.form.get('task_input'))

    logger.debug("Upload parameters - fMRI type: %s, Timecourse: %s, Task: %s", 
                    fmri_file_type, timecourse_input, task_input)
    
    # initialiize FileUpload class
    file_upload = FileUpload(
        fmri_file_type, 
        ts_status=timecourse_input, 
        task_status=task_input,
        method='browser'
    )
    logger.info("FileUpload instance initialized")

    # get file uploads
    try:
        uploads = file_upload.upload()
        logger.info("Files successfully uploaded and validated")
    except (
        exception.FileInputError,
        exception.FileUploadError,
        exception.FileValidationError
    ) as e:
        logger.error("File upload error: %s", str(e), exc_info=True)
        # if error in time course upload, pass index of failed time course file
        if e.file_type == exception.ExceptionFileTypes.TIMECOURSE.value:
            index = e.index
        else:
            index = None

        return make_response({
            "error": e.message,
            "file_type": e.file_type,
            "fields": e.field,
            "index": index
        }, 400)
    except Exception as e:
        logger.critical("Unexpected error during file upload: %s", str(e), exc_info=True)
        return make_response({
            "error": "An unexpected error occurred during file upload",
            "file_type": None,
            "fields": None,
            "index": None
        }, 500)

    # pass fmri data to data manager and get viewer data
    # Initialize state in the active context
    if fmri_file_type == 'nifti':
        data_manager.ctx.create_nifti_state(
            func_img = uploads['nifti'][NiftiFiles.FUNC.value],
            anat_img = uploads['nifti'][NiftiFiles.ANAT.value],
            mask_img = uploads['nifti'][NiftiFiles.MASK.value]
        )
        logger.info("Nifti data manager state created successfully")
    else:
        data_manager.ctx.create_gifti_state(
            left_func_img=uploads['gifti'][GiftiFiles.LEFT_FUNC.value],
            right_func_img=uploads['gifti'][GiftiFiles.RIGHT_FUNC.value],
            left_mesh=uploads['gifti'][GiftiFiles.LEFT_MESH.value],
            right_mesh=uploads['gifti'][GiftiFiles.RIGHT_MESH.value]
        )
        logger.info("Gifti data manager state created successfully")
    # initialize data manager with or without time series data
    data_manager.ctx.add_timeseries(uploads['ts'])
    if file_upload.ts_status:
        logger.info("Time series data added to viewer data")
    else:
        logger.info("No time series data added to viewer data")

    # if task data, add to viewer data
    if file_upload.task_status:
        data_manager.ctx.add_task_design(
            task_data=uploads['task']['task_regressors'], 
            tr=uploads['task']['tr'], 
            slicetime_ref=uploads['task']['slicetime_ref']
        )
        logger.info("Task design data added to viewer data")
    else:
        logger.info("No task design data added to viewer data")

    # get viewer metadata
    viewer_metadata = data_manager.ctx.get_viewer_metadata()
    logger.info("Viewer metadata retrieved successfully")
    
    return make_response(
        viewer_metadata,
        201
    )


@file_bp.route(Routes.UPLOAD_SCENE.value, methods=['POST'])
def upload_scene() -> dict:
    """Upload a scene file"""
    logger.info("Uploading scene file")
    # Check if a file was uploaded
    if 'scene_file' not in request.files:
        logger.error("No scene file provided")
        return make_response({"error": "No scene file provided"}, 400)
    
    scene_file = request.files['scene_file']
    
    # Check if file is empty
    if scene_file.filename == '':
        logger.error("Empty file provided")
        return make_response({"error": "Empty file provided"}, 400)
    
    # Validate file extension
    if not scene_file.filename.endswith('.fvstate'):
        logger.error("Invalid file format. Expected .fvstate file")
        return make_response({"error": "Invalid file format. Expected .fvstate file"}, 400)
    
    try:
        # Read the file data
        scene_file_data = scene_file.read()
        
        # Load the state
        data_manager.load(scene_file_data)
        
        # Get viewer metadata for the restored state
        viewer_metadata = data_manager.ctx.get_viewer_metadata()
        
        return make_response(viewer_metadata, 200)
    except exception.FVStateVersionIncompatibleError as e:
        logger.error(f"Error loading scene file: {str(e)}")
        return make_response(
            {"error": f"Failed to load scene: {str(e)}"}, 400)
    except Exception as e:
        logger.error(f"Error loading scene file: {str(e)}")
        return make_response({"error": f"Failed to load scene: {str(e)}"}, 500)
    

